### PR TITLE
performance improvement using jpa example

### PR DESCRIPTION
**Pull request by nahguam**
# stream-registry PR

performance improvement using jpa example

# PR Checklist Forms

- [ ] CHANGELOG.md updated
- [ ] Reviewer assigned
- [ ] PR assigned (presumably to submitter)
- [ ] Labels added (enhancement, bug, documentation)

### CODE REVIEW

File: core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/services/ProducerBindingService.java [hunk 1]
```diff
@@ -79,14 +80,25 @@ public void delete(ProducerBinding producerBinding) {
...
+    ProducerBinding example = new ProducerBinding(
+        new ProducerBindingKey(
+            key.getStreamDomain(),
+            key.getStreamName(),
+            key.getStreamVersion(),
+            key.getZone(),
+            null,
+            key.getName()
+        ), null, null);
+    return producerBindingRepository.findAll(Example.of(example)).stream().findFirst();
```
Comment (nahguam) [inline core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/services/ProducerBindingService.java]: This is where the magic happens.

Comment (rshotels) [reply core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/services/ProducerBindingService.java]: So this produces an actual HQL/ SQL query which matches the example, right?

File: core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/services/ProducerBindingService.java [hunk 2]
```diff
@@ -79,14 +80,25 @@ public void delete(ProducerBinding producerBinding) {
   }
 
   public Iterable<ProducerBinding> findAll(Predicate<ProducerBinding> filter) {
-    return stream(producerBindingRepository.findAll().spliterator(), false)
-        .filter(r -> filter.test(r))
-        .collect(Collectors.toList());
+    return producerBindingRepository.findAll().stream().filter(filter).collect(toList());
```
Comment (rshotels) [inline core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/services/ProducerBindingService.java]: Might be possible to use the 'Example' process if the filter has any non-regex fields, then apply the Predicate for the regex. Something for tomorrow?

File: model/src/main/java/com/expediagroup/streamplatform/streamregistry/model/ConsumerBinding.java [hunk 3]
```diff
@@ -36,7 +40,5 @@
   @EmbeddedId
   private ConsumerBindingKey key;
   private Specification specification;
-  private String consumerId; //Resolves to Consumer
-  private String bindingId; //Resolves to StreamBinding
```
Comment (rshotels) [inline model/src/main/java/com/expediagroup/streamplatform/streamregistry/model/ConsumerBinding.java]: Are you sure these can come out?

File: core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/services/ProducerBindingService.java (hunk 1 above)
Comment (nahguam) [reply core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/services/ProducerBindingService.java]: I imagine so.

File: core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/repositories/CachingJpaRepository.java [hunk 4]
```diff
@@ -15,26 +15,31 @@
...
 
-  @QueryHints(@QueryHint(name = org.hibernate.annotations.QueryHints.CACHEABLE, value = "true"))
-  boolean existsById(ID var1);
+  @QueryHints(@QueryHint(name = CACHEABLE, value = "true"))
+  List<T> findAll();
 
-  @QueryHints(@QueryHint(name = org.hibernate.annotations.QueryHints.CACHEABLE, value = "true"))
-  Iterable<T> findAll();
+  @QueryHints(@QueryHint(name = CACHEABLE, value = "true"))
+  List<T> findAllById(Iterable<ID> ids);
```
[the remark this answers is not in the document]
Comment (nahguam) [reply core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/repositories/CachingJpaRepository.java]: I just changed the param name to the same as it is in the super interface.

File: core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/services/ProducerBindingService.java (hunk 2 above)
Comment (nahguam) [reply core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/services/ProducerBindingService.java]: Everything is a regex string apart from the stream version which is an integer. May be worth exploring but it's not low hanging fruit and also separately from this PR.

File: model/src/main/java/com/expediagroup/streamplatform/streamregistry/model/ConsumerBinding.java (hunk 3 above)
Comment (nahguam) [reply model/src/main/java/com/expediagroup/streamplatform/streamregistry/model/ConsumerBinding.java]: They're not there on the `ProducerBindingKey` and that works, and the build tests pass 👍

Comment (rshotels) [reply model/src/main/java/com/expediagroup/streamplatform/streamregistry/model/ConsumerBinding.java]: ok

File: core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/services/ProducerBindingService.java (hunk 2 above)
Comment (rshotels) [reply core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/services/ProducerBindingService.java]: Just putting it out there before I forget.

File: core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/repositories/CachingJpaRepository.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (rshotels) [reply core/src/main/java/com/expediagroup/streamplatform/streamregistry/core/repositories/CachingJpaRepository.java]: Well ok then

**Pull request merged**